File: BOJ/seulah/5_week/햄버거 분배.py
import sys
input = sys.stdin.readline

N,K = map(int,input().split())
ham = list(input().strip())
cnt =0

for i in range(N):
    if ham[i] =="P":
        left = i-K
        right = i+K+1

        if left < 0:  # 왼쪽 범위가 0보다 작으면 0
            left = 0
        if right > N:  # 오른쪽 범위가 N을 초과하면 N으로 설정
            right = N

        for j in range(left,right):
            if ham[j] == "H":
                ham[j] = "1"
                break

print(ham.count("1"))

File: BOJ/jaehong/5_week/햄버거 분배.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];

rl.on("line", (line) => {
  input.push(line);

  if (input.length === 2) {
    const [N, K] = input[0].split(" ").map(Number);
    const arr = input[1].split("");

    // console.log(N,K,arr);

    console.log(solution(N, K, arr));
    rl.close();
  }
});

function solution(N, K, arr) {
  let people = [];
  let burgers = [];

  for (let i = 0; i < N; i++) {
    if (arr[i] === 'P') {
      people.push(i);
    } else if (arr[i] === 'H') {
      burgers.push(i);
    }
  }

  let count = 0;
  let pIndex = 0, hIndex = 0;

  while (pIndex < people.length && hIndex < burgers.length) {
    if (Math.abs(people[pIndex] - burgers[hIndex]) <= K) {
      count++;
      pIndex++;
      hIndex++;
    } else if (people[pIndex] < burgers[hIndex]) {
      pIndex++;
    } else {
      hIndex++;
    }
  }

  return count;
}
